### ISSUE TITLE
HTTP API request errors

### ISSUE THREAD

**Issue by TuguldurJ** [postalserver/postal#2242]
related issues #2231 #2237

HTTP API request response with errors below:
```
400: {"status":"invalid-controller-or-action"}
500: {"status":"internal-server-error","error":"Moonrope::Errors::MissingAuthenticator",
"message":"Wanted to use authenticator that was not defined.",
"backtrace":["/usr/local/bundle/gems/moonrope-2.0.2/lib/moonrope/request.rb:110:in `execute'",
"/usr/local/bundle/gems/moonrope-2.0.2/lib/moonrope/rack_middleware.rb:94:in `call'",
"/usr/local/bundle/gems/moonrope-2.0.2/lib/moonrope/doc_server.rb:118:in `call'",
"/opt/postal/app/lib/postal/tracking_middleware.rb:12:in `call'",
"/usr/local/bundle/gems/rack-2.2.3/lib/rack/tempfile_reaper.rb:15:in `call'",
"/usr/local/bundle/gems/rack-2.2.3/lib/rack/etag.rb:27:in `call'"]}
response: null
```

**Comment by TuguldurJ** [comment]
Could you please help this issue?
@adamcooke  [https://github.com/adamcooke/moonrope](url)

**Comment by willpower232** [comment]
> invalid-controller-or-action

there must be something wrong with the request you're making, are you using an API library or are you making the requests manually?

**Comment by TuguldurJ** [comment]
I used many API libraries, they all respond to this error. Last time I used Webflux for Spring Boot and server replies:
400 Bad POST request https://$host/api/v1/send/message
500 POST internal server error https://$host/api/v1/send/message

Almost 10% of messages receive that error.

**Comment by willpower232** [comment]
I was referring to the specific libraries to send messages with Postal as listed here https://docs.postalserver.io/developer/client-libraries 

Are you able to find anything in common in the messages that don't work?

**Comment by TuguldurJ** [comment]
1. Yes, I already tried these mentioned client libraries. @matthewmgamble which is well written in the java client library. I am currently using this library. I've also tried possible http request libraries for Java from the best repositories.

2. If I set a wrong header, all requests will fail. But its not. In every 1-2 requests out of 10 it would cause this error. 
As you can see, the problem is that when you send a request with standard header and authentication, the
postal server cannot handle asynchronous requests, which means that it cannot handle many requests (2-3 requests per sec) at once. I set the database settings to maximum availability where mentioned settings in #2237 .

Most of the people use the postal service by smtp. but if you have a dedicated web server and database, you must need the API option. Then when you send 10'000-100'000 asynchronous http requests with 16 threads it will cause error.

**Comment by willpower232** [comment]
Thanks for the extra information.

Postal does actually include a config file for Puma here https://github.com/postalserver/postal/blob/main/config/puma.rb

I casually saw a message on mastodon which also seems to be useful https://ruby.social/@evanphx/109365852402052564

tl;dr add `workers` to the config file with a value of 1.5 the CPU core count available and then adjust threads and see what happens, I can't test that this genuinely helps but I don't know what else to suggest right now. The developer warns that high thread values will see diminishing returns so need to be aware of that as well

**Comment by TuguldurJ** [comment]
Thank you @willpower232!
I changed the puma config and enabled cluster mode. And now our web server can handle async API requests.
In this configuration we can handle a maximum of 4096 API requests. 

```
require_relative '../lib/postal/config'
WEB_CONCURRENCY = 32
workers 32 # enables cluster mode
threads_count   = 128 # 1 worker per 128 threads
#threads_count  = Postal.config.web_server&.max_threads&.to_i || 5
threads         threads_count, threads_count
bind_address    = Postal.config.web_server&.bind_address || '127.0.0.1'
bind_port       = Postal.config.web_server&.port&.to_i || ENV['PORT'] || 5000
bind            "tcp://#{bind_address}:#{bind_port}"
environment 'application'
#environment     Postal.config.rails&.environment || 'development'
prune_bundler
quiet false
preload_app!
```

**Comment by willpower232** [comment]
good stuff, I would probably restore the `environment` and `threads_count` lines and then add 128 to the config file.

I don't think you need the WEB_CONCURRENCY line as that is supposed to be an environment variable when running puma directly.

**Comment by TuguldurJ** [comment]
Yeah, that would be fine. 

I also found out that the 400 error has 2 possible reasons:
1. Puma server cannot handle bulk async requests with single mode. Thus, we enabled cluster mode it by above the config and it works.
2. If you have poor network connection between Mta server and Backend server. It will cause packet loss and 400 error.
So I think that API requests should be secure and with a good connection without any packet loss. max rtt should be under 1ms like this below between 2 servers:
```
rtt min/avg/max/mdev = 0.299/0.322/0.335/0.012 ms
```
I configured 2 servers as vRack (postal mta server, backend server) with private network. 
```
Backend server: 192.168.0.1
Postal MTA server: 192.168.0.2
```
Now I have questions regarding the configuration:
1. How can we send API request by private IP address like this: https://192.168.0.2/api/v1/send/message? I checked CaddyFile and added extra line for it.
```
192.168.0.1 {
  reverse_proxy 127.0.0.1:5000
}
```   
But Caddy cannot handle IP requests and it gives me an error:
```
tlsv1 alert internal error
```

2. What is the difference between "development" and "application" environment in puma.rb? 
3. Where should I save puma.rb permanently in docker? Currently, I temporarily saved it to /var/lib/docker/overlay2/.../diff/opt/postal/app/config/puma.rb

**Comment by willpower232** [comment]
1 - I would presume caddy is trying to treat the IP address as a domain name so you should assign a different domain name for that rather than an IP. You can't do certificate verification on an IP address so even if you exposed it, you would have to use HTTP probably

2 - I think it mostly affects the assets and error messages that the UI receives

3 - you could either build a docker image including the updated file or list it under the config volume entries in the docker-compose.yml

**Comment by TuguldurJ** [comment]
1. Yes, we may need to use a different domain name for the IP addresses.
2. As a long time stress test and debugging, I finally fixed this problem. Which HTTP client service you use, you should retry the request if it fails and takes time. In Java, use the webflux reactive spring http client with the following code: `retryWhen(Retry.fixedDelay(RETRY_DELAY_TIME, Duration.ofMinutes(
                                                  1)))`
3. Finally, I built my own docker image including updated file which is fine.
Thanks, for those who might have scalability performance issues. this session would help in the future.
